### DISCUSSION TITLE
Nats Request/Reply works in Debug but error when run standalone

### DISCUSSION THREAD

**Discussion by alberk8** [nats-io/nats.net#468] [Q&A] [answered]
I have a Nats Request/Reply in two apps.   On the Reply app on VS debug there is no issue but when I compile and run it in CLI (myapp.exe) 

I am getting the error below after the first Request/Reply call.   What could be the issue?

`Subscription GCd but was never disposed`

Another message that I see is sometimes on Request/Reply is  `End subscription MaxMsgs`.  What does this mean?

**Comment by mtmk** [comment]
we have some protection against subscriptions that are not properly disposed which involves weak refs / GC etc. Can you post a Program.cs (ideally, or some code) demonstrating the issue?

**Reply by alberk8** [reply]
I have a IHostedService and when the _svc, _dataService and _group is declared as local (ie var _svc).

```cs
public async Task StartAsync(CancellationToken cancellationToken)
{
     try
     {
         var _svc = new NatsSvcContext(_natsConnection);

         var _dataService = await _svc.AddServiceAsync("systemdata", "1.0.0");
         var _group = await _dataService.AddGroupAsync("systemgroup, "systemQ");
         
         await _group.AddEndpointAsync<byte[]>(GetSystemInfo, "getperson");

         
        
         Debug.WriteLine("Start NATs Service");
     }
     catch (Exception ex)
     {
         _logger.LogInformation($"Nats Hosted Services Error: {ex.Message}");
     }

     return;
}

public static async ValueTask GetSystemInfo(NatsSvcMsg<byte[]> data)
{

    try
    {
        // Get data and reply
        await data.ReplyAsync();
    }
    catch (NatsException ex)
    {
        await data.ReplyErrorAsync(999, ex.Message);
    }
    catch (Exception ex)
    {
        await data.ReplyErrorAsync(500, ex.Message);
    }
}
```

**Reply by mtmk** [reply]
haven't tried the code yet, but first thing I'd suggest is to make them class members. Otherwise, I think they'd be garbage collected and subscriptions would go away.

**Reply by alberk8** [reply]
I did the following  and it seems to work.

```cs
public class NatsHostedService : IHostedService, IAsyncDisposable
{
 private NatsSvcContext _svc;
 private INatsSvcServer _dataService;
 private NatsSvcServer.Group _group;
```

**Reply by AngeloBestetti** [reply]
I have the same problem, but I was creating static classes, I changed now and implement IAsyncDisposable. Waiting now to see if happens again.

Thank you for the tip

**Reply by AngeloBestetti** [reply]
Didn´t work, still getting the same message

**Reply by mtmk** [reply]
could you post a Program.cs for me to reproduce the issue?

**Reply by AngeloBestetti** [reply]
`using NATSServicesSOAWebServices.Services;

namespace NATSServicesSOAWebServices
{
	internal class Program
	{
		static async Task Main(string[] args)
		{
			RegisterServices rs = new RegisterServices();
			await rs.NATSServices();
		}
	}
}
`

**Reply by AngeloBestetti** [reply]
`using Microsoft.Extensions.Logging;
using NATS.Client.Core;
using NATS.Client.JetStream;
using NATS.Client.Serializers.Json;
using NATS.Client.Services;
using NATSServicesSOAWebServices.Models;
using NATSServicesSOAWebServices.Models.Administracao;
using Newtonsoft.Json.Schema;
using Newtonsoft.Json.Schema.Generation;

namespace NATSServicesSOAWebServices.Services
{
	public class RegisterServices : IAsyncDisposable
	{
		internal static NatsConnection natsConnection { get; set; }
		internal static NatsJSContext natsJSContext { get; set; }

		internal static string nats_servers = "list of servers";

		internal static JSchemaGenerator schemaGenerator = new JSchemaGenerator();

		public async Task NATSServices()
		{

			using var loggerFactory = LoggerFactory.Create(builder => builder.AddConsole());
			var logger = loggerFactory.CreateLogger("NATS-PessoaFisicaNFe");
			var url = $"{nats_servers}";
			var opts = new NatsOpts
			{
				Url = url,
				SerializerRegistry = NatsJsonSerializerRegistry.Default,
				Name = $"NATS-{Environment.MachineName}",
				IgnoreAuthErrorAbort = true,
				LoggerFactory = loggerFactory,
				AuthOpts = new NatsAuthOpts()
				{
					Username = "username",
					Password = "password"
				},
				RequestTimeout = TimeSpan.FromSeconds(30),
				ConnectTimeout = TimeSpan.FromSeconds(30),
				CommandTimeout = TimeSpan.FromSeconds(30),
				SubscriptionCleanUpInterval = TimeSpan.FromSeconds(30),
			};
			await using NatsConnection nats = new NatsConnection(opts);

			nats.ConnectionOpened += Utils.Nats_ConnectionOpened;
			nats.ConnectionDisconnected += Utils.Nats_ConnectionDisconnected;

			await nats.ConnectAsync();


			// Jetstream
			var js = new NatsJSContext(nats);
			var svcNatsContext = new NatsSvcContext(nats);

		


			try
			{
				Dictionary<string, string> metadata = new Dictionary<string, string>();
				metadata.Add("Origem", "RFB - Receita Federal do Brasil");
				metadata.Add("Transacao", "Pessoa Fisica");
				metadata.Add("Descricao", "Consulta de Pessoa Fisica na Receita Federal do Brasil");
				metadata.Add("Ambiente", Environment.MachineName);
				var serviceRFBPessoaFisica = await svcNatsContext.AddServiceAsync(new NatsSvcConfig("PessoaFisicaNFe", "1.0.2")
				{
					Description = "Consulta RFB Pessoa Fisica NFe",
					Metadata = metadata,
				}, cancellationToken: CancellationToken.None);

				NATSServicesSOAWebServices.Services.RFB.PF pf = new NATSServicesSOAWebServices.Services.RFB.PF(nats, js);
				await serviceRFBPessoaFisica.AddEndpointAsync(pf.GetPessoaFisicaNFe, "PessoaFisicaNFe", serializer: NatsJsonSerializer<PessoaFisicaNFeRequisicao>.Default);
			}
			catch (Exception ex)
			{
				Console.WriteLine(ex.Message);
			}


			try
			{
				while (true)
				{
					System.Threading.Thread.Sleep(300);
				}
			}
			catch (NatsException ex)
			{
				Console.WriteLine(ex.Message);
			}
			catch (Exception ex)
			{
				Console.WriteLine(ex.Message);
			}
		}

		public ValueTask DisposeAsync()
		{
			natsConnection.DisposeAsync();
			return ValueTask.CompletedTask;
		}
	}
}
`

**Reply by mtmk** [reply]
could you try making rs a class field?

edit: I actually meant a self contained Program.cs so I can run it on my machine 😄 but let's see if above works.

**Reply by AngeloBestetti** [reply]
I put everything inside the Program.cs and deployed, let´s see if works

**Reply by AngeloBestetti** [reply]
same thing, after 1 hour I´m receiving same message

**Reply by darkwatchuk** [reply]
How are you calling the method NATSServices, are you awaiting it?   
Also, even though you have (static for some reason?) natsConnection and natsJSContext  variables, they're not being used, instead, you're using method variables.  Additionally if you're going to wait inside this method rather than outside, probably best to use Task.Delay than Thread.Sleep

**Reply by mtmk** [reply]
good questions.

also issue might be this line:
```csharp
var serviceRFBPessoaFisica = await svcNatsContext.AddServiceAsync(new NatsSvcConfig("PessoaFisicaNFe", "1.0.2")
```
try with `serviceRFBPessoaFisica` being a class field. that variable would be the one holding references to subscriptions under the hood.

**Reply by AngeloBestetti** [reply]
Didn´t work, I placed GC.Collect() to force the garbage collection, let´s see in an hour if happens again

**Reply by mtmk** [reply]
not sure what's happening tbh. my suggestion is same as @darkwatchuk said try to get rid of statics and thread.sleep etc. if it still fails, can you create a minimal self-contained project I can try to run it on my machine and debug on my end.

**Reply by AngeloBestetti** [reply]
GC.Collect() Works fine, 4 hours working didn´t die

**Reply by mtmk** [reply]
@AngeloBestetti I'm not convinced running GC.Collect() is helping here. Did you have any progress on this?

**Reply by AngeloBestetti** [reply]
I wrote in go and works fine now

**Reply by mtmk** [reply]
glad you solved the issue. sad to see you go (no pun intended). let us know if you decide to use C# again.

**Reply by AngeloBestetti** [reply]
@mtmk Ziya, found something strange, when I use just reply msg works and does not lost the "subscription", what is happening is my entity framework, before I reply, I need to query a DB to get some info. If I remove this part there is no error. If I put back after few seconds I got the Message.

Does EF Core interfere with NATS?

`
ClienteModel cliente = await Utils.GetCliente(req.Credenciais.Email, req.Credenciais.Senha);     <------ This is the Problem, if I remove no errors
resposta = await Utils.GetPessoaFisicaNFe(req);
await msg.ReplyAsync(resposta);
Console.WriteLine($"Nome: {JsonConvert.SerializeObject(resposta)}");
`

**Reply by mtmk** [reply]
thanks @AngeloBestetti. this is interesting! I think it might be to do with `await` creating the state machine code behind the scenes. @caleblloyd has a fix in #506. have a look at [Kevin Gosse's post](https://minidump.net/c-using-gc-keepalive-in-async-methods-8d20fd79f0a0/) he's mentioned. I think that blog post might explain the behaviour you're seeing above.

* #506

**Comment by darkwatchuk** [comment]
I've previously found if you don't await / pin a long running task then it gets GC'd

**Reply by mtmk** [reply]
@darkwatchuk fyi, I thought Kevin Gosse's post @caleblloyd found (see #506) was quite interesting: https://minidump.net/c-using-gc-keepalive-in-async-methods-8d20fd79f0a0/

**Reply by mtmk** [reply]
Hi @AngeloBestetti, we have a new release that should fix the issue you mentioned above. Please give it a go and let us know if it fixes the issue for you, cheers: https://github.com/nats-io/nats.net.v2/releases/tag/v2.2.3

**Comment by mtmk** [comment]
Hi @alberk8 https://github.com/nats-io/nats.net.v2/releases/tag/v2.2.3 should fix this issue

**Reply by AngeloBestetti** [reply]
@mtmk Great, I just deployed and tested, WORKS fine, let´s run for few hours. Thank you

**Reply by AngeloBestetti** [reply]
Working perfect now, thank you Ziya

**Reply by alberk8** [reply, marked as answer]
After a few days, still running.

**Reply by mtmk** [reply]
All credits to @caleblloyd for the fix #506. Thank you 🙏
